src/Common/src/Interop/Unix/System.Native/Interop.FdSet.cs

```diff
@@ -2,34 +2,54 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Diagnostics;
 using System.Runtime.InteropServices;
 
 internal static partial class Interop
 {
     internal static partial class Sys
     {
-        public unsafe struct FdSet
-        {
-            const int FDSET_MAX_FDS = 1024;
-            const int FDSET_NFD_BITS = FDSET_MAX_FDS / (8 * sizeof(uint));
+        [DllImport(Libraries.SystemNative)]
+        private static extern int FdSetSize();
+
+        private static readonly int FD_SETSIZE_BITS = FdSetSize();
+
+        private const int BitsPerByte = 8;
+        private static readonly int FD_SETSIZE_BYTES = FD_SETSIZE_BITS / BitsPerByte;
+
+        internal static readonly int FD_SETSIZE_UINTS = FD_SETSIZE_BYTES / sizeof(uint);
 
-            private fixed uint _bits[FDSET_MAX_FDS / FDSET_NFD_BITS];
-  
-            public void Set(int fd)
+        internal static unsafe void FD_SET(int fd, uint* fdset)
+        {
+            if (fd >= FD_SETSIZE_BITS)
             {
-                fixed (uint* bits = _bits)
-                {
-                    bits[fd / FDSET_NFD_BITS] |= (1u << (fd % FDSET_NFD_BITS));
-                }
+                ThrowInvalidFileDescriptor(fd);
             }
+            Debug.Assert(fdset != null);
+
+            fdset[fd / FD_SETSIZE_UINTS] |= (1u << (fd % FD_SETSIZE_UINTS));
+        }
 
-            public bool IsSet(int fd)
+        internal static unsafe bool FD_ISSET(int fd, uint* fdset)
+        {
+            if (fd >= FD_SETSIZE_BITS)
             {
-                fixed (uint* bits = _bits)
-                {
-                    return (bits[fd / FDSET_NFD_BITS] & (1U << (fd % FDSET_NFD_BITS))) != 0;
-                }
+                ThrowInvalidFileDescriptor(fd);
             }
+            Debug.Assert(fdset != null);
+
+            return (fdset[fd / FD_SETSIZE_UINTS] & (1u << (fd % FD_SETSIZE_UINTS))) != 0;
+        }
+
+        internal static unsafe void FD_ZERO(uint* fdset)
+        {
+            Debug.Assert(fdset != null);
+            Interop.Sys.MemSet(fdset, 0, (UIntPtr)FD_SETSIZE_BYTES);
```

Comment:
Simply because it was measurably faster, in particular when FD_SETSIZE is set to 65K, at which point the loop would need to iterate 2K times.  If we had an Buffer.MemorySet (parallel to Buffer.MemoryCopy), I'd have happily used that.

If you think the perf here doesn't matter, I can change it back to being a loop, which is actually what I initially had.